src/docs/arch-concurrency.dox

```diff
@@ -0,0 +1,110 @@
+/*! @arch_page arch-concurrency WiredTiger concurrency management
+
+WiredTiger is a multi-threaded application and needs to manage concurrent access to shared memory
+locations. Many mechanisms are defined in WiredTiger to safely access shared memory locations. This
+page describes what those mechanisms are and the intended usage of them in the code base.
+
+@section volatile Volatile
+
+A controversial tool in the C toolkit. WiredTiger utilizes volatile for one reason, managing
+compiler optimization. Leveraging the definitions from "Is Parallel Programming Hard, And, If So,
+What Can You Do About It?", Paul E. McKenney, WiredTiger utilizes volatile to prevent the
+following type of compiler optimizations: Load fusing, store fusing, invented loads, invented
+stores and code reordering. WiredTiger defines the following macros WT_READ_ONCE and
+WT_WRITE_ONCE, these macros, if used correctly, prevent the described forms of compiler
+optimization. Additionally some variables in WiredTiger are defined with the volatile keyword if
```

Comment:
Comment from Chenhao: Can we explain why we need to define the frequently used concurrent variables as volatile?

Reply:
No I cannot, personally I don't think we should define variables as volatile and instead use read_once write_once. But that would be too large a change and so I put this sentence in to manage that.